code/Control/UserDefinedFormController.php

```diff
@@ -511,6 +528,152 @@ public function finished()
+            $editableFormField->setClassName(EditableFileField::class);
+            $editableFormField->write();
+        }
+        $treeView->setSchemaData([
+            'data' => [
+                'urlTree' => "admin/pages/edit/EditForm/$userFormID/field/Fields/item/$id/ItemEditForm/field/FolderID/tree"
+            ]
+        ]);
+
+
+        // create the schema response
+        $parts = $this->getRequest()->getHeader(LeftAndMain::SCHEMA_HEADER);
+        $schemaID = $this->getRequest()->getURL();
+        $data = FormSchema::singleton()
+            ->getMultipartSchema($parts, $schemaID, $form);
+
+        // return the schema response
+        $response = new HTTPResponse(json_encode($data));
+        $response->addHeader('Content-Type', 'application/json');
+        return $response;
+    }
+
+    /**
+     * Sets the selected folder as the upload folder for an EditableFileField
+     * @return HTTPResponse
+     * @throws ValidationException
+     */
+    public function confirmfolder()
+    {
+        $request = $this->getRequest();
+
+        // check if we're creating a new folder or using an existing folder
+        $option = $request->requestVar('FolderOptions');
+        if ($option == 'existing') {
+            // set existing folder
+            $folderID = $request->requestVar('FolderID');
+            $folder = Folder::get()->byID($folderID);
+        } else {
+            // create the folder
+            $createFolder = 'SecureUploads/' . $request->requestVar('CreateFolder');
+            $folder = Folder::find_or_make($createFolder);
+
+            // ensure the SecureUploads folder is only accessible to Administrators
+            $secureUploadsFolder = Folder::find("SecureUploads");
+            $secureUploadsFolder->CanViewType = InheritedPermissions::ONLY_THESE_USERS;
+            $secureUploadsFolder->ViewerGroups()->removeAll();
+            $secureUploadsFolder->ViewerGroups()->add(Group::get_one(Group::class, ['Code' => 'administrators']));
+            $secureUploadsFolder->write();
+        }
+
+        // retrieve the EditableFileField
+        // change the class if it is incorrect
+        $id = $request->requestVar('ID');
+        /** @var EditableFileField $editableFileField */
+        $editableFormField = EditableFormField::get()->byID($id);
+        if (!$editableFormField) {
+            $editableFormField = Versioned::get_by_stage(EditableFormField::class, Versioned::DRAFT)->byID($id);
+        }
+        if (!$editableFormField instanceof EditableFileField) {
+            $editableFormField->setClassName(EditableFileField::class);
+        }
+        $editableFileField = $editableFormField;
+
+        // assign the folder
+        $editableFileField->Folder = $folder;
+        $editableFileField->write();
+
+        // respond
+        $response = new HTTPResponse(json_encode([]));
+        $response->addHeader('Content-Type', 'application/json');
+        return $response;
```

Comment:
Wouldn't it catch the error and return a different response in that case?

Reply:
It'll return a big blob of HTML, which you won't be able to use to display a useful error. Basically I'm suggesting wrapping this logic in additional checks to ensure all the parameters are present, etc., and then responding with JSON if there are any failures, rather than waiting for core to blurt out a default error response.